onnxscript/ir/_name_authority.py

```diff
@@ -12,20 +12,53 @@ class NameAuthority:
     ``node_{op_type}_{node_counter}`` for nodes. The counter is incremented each time
     a new value or node is named.
 
-    The class does not keep track of the names it has given, so it is possible to
-    generate names that conflicts with existing names. It is the responsibility of the
-    user to ensure that the names are unique (typically by running a name-fixing pass
-    on the graph).
+    This class does keep track of the names it has given and all the existing names
+    in the graph to prevent producing duplicated names.
+
+    However, if a value/node is already named when added to the graph,
+    the name authority will not change its name.
+    It is the responsibility of the user to ensure that the names are unique
+    (typically by running a name-fixing pass on the graph).
+    TODO(justichuby): Describe the pass when we have a reference implementation.
     """
 
     def __init__(self):
         self._value_counter = 0
         self._node_counter = 0
+        self._value_names: set[str] = set()
+        self._node_names: set[str] = set()
+
+    def _unique_value_name(self) -> str:
+        """Generate a unique name for a value."""
```

Comment:
Would it make sense to allow callers to specify a desired name, which can be used as a prefix (instead of `val`)? Or, does that functionality reside elsewhere? That is, if we have rewrite-rule that introduces names like `shape` and `dim`, how do we turn them into `shape_102` instead of `val_102`?

Reply:
Currently it’s not handled. Happy to explore options